[FIX] l10n_it_stock_ddt: support dropshipping in ddt

agr-odoo opened the pull request:
Description of the bug:
- Have an IT company configured
- Activate Dropship
- Create a Product P with dropship enabled and vendor configured
- Create a quotation to an IT customer
- Add P to a line
- Confirm
- Purchase will be created automatically
- Confirm it.
- Go to dropship picking
- Confirm
- Print

The DDT report does not show up correctly:
- Warehouse address is the company address
- Customer address is the vendor address

Fast link to video https://watch.screencastify.com/v/9Do2xXWmnwx8pHuFFtck
https://www.odoo.com/web#id=3128812&model=project.task
opw-3128812

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/110398).

agr-odoo commented:
@jco-odoo

oco-odoo commented:
Shouldn't this rather be reviewed by @odoo/rd-logistics  ?

jco-odoo commented:
@oco-odoo We need the info in the invoice.  Without this, the feature does not make any sense here anyways.

lordkrandel commented:
cc @jco-odoo

The PR is generally good, but there's something to be added.

Well, this was very hard to test since the action `Print` -> `DDT Report` which I suppose it's custom or from a later version, and the conditions on making the correct `Print` button appear weren't all that clear to me.
In detail, the `picking_type_id.code` set to `incoming` for Dropshipping doesn't really make sense (but whatever).
![image](https://user-images.githubusercontent.com/1665365/234326686-092c505d-7904-462f-a009-5a077a25e9db.png)

Fact is, the DDT should be printed by the normal `Print` button of the `stock.picking` as it is for non-dropshipping orders. So I made a computed field to allow that case to be handled, and added the commit here.

We should add a test in  `odoo/addons/l10n_it_stock_ddt/tests/test_ddt.py` to test that the dropshipping address stays correct, so that somehow the behaviour is documented and updated in case of change.
Please, add the test, review and squash if needed. I'd do this myself but will be off for some time.
This gist here (the ending part about dropshipping) can guide you a bit maybe.
      https://gist.github.com/lordkrandel/ea3f35c4c4b0c417079be5781353f0b7

Thank you and sorry for the wait!

agr-odoo commented:
@lordkrandel Thanks for the addition and the info

Added a test and rebased. As we have two different fixes I would keep them in separate commits

My only concern is with the added test: dropship is defined in purchase, but we don't have a bridge module with the ddt functionality. Is this acceptable or are there better workarounds?
https://github.com/odoo/odoo/pull/110398/files#diff-38e1e5ced8491ffc8512ea0337cda3ab0a2ce236ebd7edef536999196978d0a0R154

cc @jco-odoo

agr-odoo commented:
> * Make sure that the test is run on the runbot, or it's mostly useless.

To fulfill this point, I had to move the test to `stock_dropshipping`, as we don't execute l10n_ tests in ci/runbot
So the check for the installed module is not necessary anymore

agr-odoo commented on addons/l10n_it_stock_ddt/models/stock_picking.py:
```diff
@@ -22,6 +22,29 @@ class StockPicking(models.Model):
+                and picking.state == 'done'
+                and picking.is_locked
+                and (picking.picking_type_code == 'outgoing'
+                     or (picking.picking_type_code == 'incoming'
```
Done, thanks :+1:

agr-odoo commented on addons/l10n_it_stock_ddt/report/l10n_it_ddt_report.xml:
```diff
@@ -3,29 +3,34 @@
     <template id="report_ddt_view">
         <t t-call="web.external_layout">
             <t t-set="o" t-value="o.with_context(lang=lang)"/>
+            <t t-if="o.move_ids_without_package and o.move_ids_without_package[0].partner_id and o.picking_type_id.code == 'incoming' and o.location_dest_id.usage == 'customer' and o.location_id.usage == 'supplier'">
```
indeed seems unnecessary, removed

agr-odoo commented on addons/stock_dropshipping/tests/test_dropship.py:
```diff
@@ -121,3 +121,62 @@ def test_00_dropship(self):
             ('product_id', '=', drop_shop_product.id)])
         self.assertEqual(len(move_line.ids), 1, 'There should be exactly one move line')
+
+    def test_dropship_addresses_report(self):
```
> where you can change the dependencies of l10n_it_stock_ddt

All dropship logic is in `stock_dropshipping`, so even if we add the dependencies it won't help

jco-odoo replied:
@agr-odoo Checking further, I don't think it is needed even to add it as a dependency.  You could actually do the dropshipping test without installing the module by creating the stock.rule/route yourself.  But adding a generic test for in the stock_dropshipping module to fix something for Italy still does not make sense.

agr-odoo replied:
>  You could actually do the dropshipping test without installing the module by creating the stock.rule/route yourself

I already tried this: we need the `buy` action when defining the rule but it's only added in `purchase_stock` (not installed as dependency)
```python
        self.env['stock.rule'].create({
            'name': '%s → %s' % (supplier_location.name, customer_location.name),
            'action': 'buy',
            'location_id': customer_location.id,
            'location_src_id': supplier_location.id,
            'procure_method': 'make_to_stock',
            'route_id': dropshipping_route.id,
            'picking_type_id': dropship_picking_type.id,
            'company_id': self.env.company.id,
        })
```
Without the full `stock_dropshipping` + `purchase` flow the test would be trivial

> But adding a generic test for in the stock_dropshipping module to fix something for Italy still does not make sense.

I agree, the reason for the move was that a test in `l10n_it_stock_ddt` has a reduced usefulness as it is run only in the nightly builds

All considered we could leave out the test

lordkrandel replied:
I can agree... but I'll explain my point so I feel more understood 😄.

The e-commerce business in Italy is all about dropshipping DDTs so I fear this kind of relationship between modules will emerge again in the future, so that's why I'd like to have tests on this flow. We already have customers dropshipping Nespresso/Lavazza coffee bags from wholesaler to clients.
(DDTs are an underdeveloped feature at the moment in the Italian localization.)

For now, let's be lean and skip the test, but let's keep that in mind.

jco-odoo commented:
@lordkrandel @agr-odoo Please remove test

agr-odoo commented:
> Please remove test

Done :+1:

jco-odoo commented:
@robodoo r+ rebase-ff
